### PR TITLE
Improved Null-checking annotations

### PR DESCRIPTION
**Pull request by rPraml**
Hello @rbygrave, in the following example

```java
String value = query.findSingleAttribute()
if (value == null) ...
```
IntelliJ complains, that `value` could never be null, which is not true.

In this PR, I tried to improve the API and added more `@NonNullApi` / `@Nullable` annotations.

I moved some null checks, so this PR is only 99% no code change.

Please take a look, if the annotations makes sense this way.

### CODE REVIEW

File: ebean-api/src/main/java/io/ebean/DB.java [hunk 1]
```diff
@@ -251,7 +250,7 @@ public static Transaction currentTransaction() {
    * </ul>
    */
   public static void flush() {
-    currentTransaction().flush();
+    getDefault().flush();
```
Comment (rPraml) [inline ebean-api/src/main/java/io/ebean/DB.java:253]: Delegating to DefaultServer.flush, which may throw a "no currentTransaction" exception.

File: ebean-api/src/main/java/io/ebean/DB.java [hunk 2]
```diff
@@ -313,7 +312,7 @@ public static void endTransaction() {
    * Mark the current transaction as rollback only.
    */
   public static void setRollbackOnly() {
-    getDefault().currentTransaction().setRollbackOnly();
+    Objects.requireNonNull(currentTransaction(),"no currentTransaction").setRollbackOnly();
```
Comment (rPraml) [inline ebean-api/src/main/java/io/ebean/DB.java:315]: Note: There is no "setRollbackOnly" in DefaultServer. Intention?

File: ebean-api/src/main/java/io/ebean/Database.java [hunk 3]
```diff
@@ -35,7 +35,7 @@
  * <h5>The 'default' Database</h5>
  * <p>
  * One Database can be designated as the 'default' or 'primary' Database
- * (see {@link DatabaseConfig#setDefaultServer(boolean)}. Many methods on DB
+ * (see {@link DatabaseConfig#setDefaultServer(boolean)}). Many methods on DB
```
Comment (rPraml) [inline ebean-api/src/main/java/io/ebean/Database.java:38]: Fixed missing `)`

File: ebean-api/src/main/java/io/ebean/Database.java [hunk 4]
```diff
@@ -166,13 +167,15 @@ public interface Database {
   /**
    * Return the BeanState for a given entity bean.
    * <p>
-   * This will return null if the bean is not an enhanced entity bean.
+   * This will throw an IllegalArgumentException if the bean is not an enhanced entity bean.
```
Comment (rPraml) [inline ebean-api/src/main/java/io/ebean/Database.java:170]: checked impl. will never return null

File: ebean-api/src/main/java/io/ebean/ExpressionList.java [hunk 5]
```diff
@@ -404,6 +404,7 @@
    *
    * }</pre>
    */
+  @Nullable
   default <A> A findSingleAttribute() {
```
Comment (rPraml) [inline ebean-api/src/main/java/io/ebean/ExpressionList.java:408]: This was the root cause, why I created this PR

File: ebean-core/src/main/java/io/ebeaninternal/api/SpiEbeanServer.java [hunk 6]
```diff
@@ -14,7 +15,6 @@
 import io.ebeaninternal.server.query.CQuery;
 import io.ebeaninternal.server.transaction.RemoteTransactionEvent;
 
-import javax.annotation.Nullable;
```
Comment (rPraml) [inline ebean-core/src/main/java/io/ebeaninternal/api/SpiEbeanServer.java:17]: Note: Wrong annotation was used here.

File: ebean-core/src/main/java/io/ebeaninternal/server/core/DefaultServer.java [hunk 7]
```diff
@@ -586,11 +586,7 @@ public void clearQueryStatistics() {
    */
   @Override
   public <T> T createEntityBean(Class<T> type) {
-    final BeanDescriptor<T> desc = descriptor(type);
-    if (desc == null) {
-      throw new IllegalArgumentException("No bean type " + type.getName() + " registered");
-    }
-    return desc.createBean();
+    return descriptor(type).createBean();
```
Comment (rPraml) [inline ebean-core/src/main/java/io/ebeaninternal/server/core/DefaultServer.java:589]: Code change: I moved this check to `descriptor(class)` as it claims to never return null due `@NonNullApi`

File: ebean-core/src/main/java/io/ebeaninternal/server/core/DefaultServer.java [hunk 8]
```diff
@@ -630,7 +626,11 @@ private <T> T referenceFindOne(Class<T> type, Object id, BeanDescriptor<?> desc)
     }
     // we actually need to do a query because we don't know the type without the discriminator
     // value, just select the id property and discriminator column (auto added)
-    return find(type).select(idProp.name()).setId(id).findOne();
+    T bean = find(type).select(idProp.name()).setId(id).findOne();
+    if (bean == null) {
+      throw new NullPointerException("Could not find reference bean " + id + " for " + desc);
+    }
```
Comment (rPraml) [inline ebean-core/src/main/java/io/ebeaninternal/server/core/DefaultServer.java:632]: Code change: I throw an NPE now.

File: ebean-api/src/main/java/io/ebean/BeanFinder.java [hunk 9]
```diff
@@ -63,6 +64,7 @@ public Database db() {
   /**
    * Return the current transaction.
```
Comment (jnehlmeier) [inline ebean-api/src/main/java/io/ebean/BeanFinder.java:65]: ... or null if there is no current transaction in scope.

File: ebean-api/src/main/java/io/ebean/Database.java [hunk 10]
```diff
@@ -866,7 +871,7 @@ public interface Database {
    * the bean does not exist then this returns false indicating that nothing was
    * deleted. However, if JDBC batch mode is used then this always returns true.
    */
-  boolean delete(Object bean, Transaction transaction) throws OptimisticLockException;
+  boolean delete(Object bean, @Nullable Transaction transaction) throws OptimisticLockException;
```
Comment (jnehlmeier) [inline ebean-api/src/main/java/io/ebean/Database.java:874]: Really? (haven't checked it). But if JavaDoc states `... with an explicit transaction.` on an API method it feels weird to me that this explicit transaction can actually be `null`. Could very well hide some coding problems in user code.

If that is really true, then JavaDoc should describe that you can pass in `null` as a transaction and what this means if you do so. Same for all the other methods taking an explicit transaction as input parameter. In that case JavaDoc of the equivalent method without transaction parameter could say something like `this method behaves exactly as passing a null transaction to delete(bean, transaction)` and link to the other method. Would avoid duplicating JavaDoc on all these methods.

File: ebean-core/src/main/java/io/ebeaninternal/server/core/DefaultServer.java (hunk 8 above)
Comment (jnehlmeier) [reply ebean-core/src/main/java/io/ebeaninternal/server/core/DefaultServer.java:632]: NPE is more like for preconditions or if you really messed up your code, don't you think? How does this exception propagate? I would not expect to received a NPE in user code if an entity isn't found. Isn't there an EntityNotFoundException?

**Comment by jnehlmeier** [comment]
Generally I feel like Ebean JavaDoc could make more use of `@return`, `@throws` and `@param` so you don't have to read through possibly lengthly JavaDoc text to get the details.

File: ebean-core/src/main/java/io/ebeaninternal/server/core/DefaultServer.java (hunk 7 above)
Comment (rPraml) [reply ebean-core/src/main/java/io/ebeaninternal/server/core/DefaultServer.java:589]: Hmm. Unfortunaltely, this is not completely true.
While DefaultServer is annotated with `@NonNullApi`, `SpiEbeanServer` does not. And we have also some code, in our app, where we check if a class is registered. So it  might be better to annotate descriptor with @Nullable

File: ebean-core/src/main/java/io/ebeaninternal/server/core/DefaultServer.java (hunk 8 above)
Comment (rPraml) [reply ebean-core/src/main/java/io/ebeaninternal/server/core/DefaultServer.java:632]: yes, thanks. EntityNotFoundException is the better choice here

File: ebean-api/src/main/java/io/ebean/Database.java (hunk 10 above)
Comment (rPraml) [reply ebean-api/src/main/java/io/ebean/Database.java:874]: > Really? (haven't checked it). But if JavaDoc states ... with an explicit transaction.

Yes, the implementation checks it, and the methods without transaction parameter just pass null there.

> If that is really true, then JavaDoc should describe that you can pass in null as a transaction and what this means if you do so. 

ACK

**Comment by rPraml** [comment]
> Generally I feel like Ebean JavaDoc could make more use of @return, @throws and @param so you don't have to read through possibly lengthly JavaDoc text to get the details.

I partially agree, What I don't like, if you just repeat the documentation like in this example ¹
```java
/**
 * Returns the current transaction
 * @return the current transaction
 */ 
 Transaction currentTransaction();
```
But you are right, for longer javadocs, a small summary would not too bad.

(I don't want to (have to) revise the complete documentation in this PR, I just want to fix the NPE warnings in the IDE ;) )

¹) Personally, I read the documentation in 90% in the IDE by jumping to the source. (Or in doubt, reading the source itself, because code never lies, while documentation can be incorrect)

**Comment by rob-bygrave** [comment]
Right so I will not merge this PR because it contains some changes that I don't like / see as being a negative.

Some changes look good but once they are included with other changes that went into behaviour changes and other changes that I believe are net negative.

File: ebean-api/src/main/java/io/ebean/Database.java [hunk 11]
```diff
@@ -120,6 +120,7 @@ public interface Database {
   /**
    * Return the associated read only DataSource for this Database instance (can be null).
    */
+  @Nullable
```
Comment (rob-bygrave) [inline ebean-api/src/main/java/io/ebean/Database.java:123]: I suspect net negative

File: ebean-api/src/main/java/io/ebean/Database.java (hunk 4 above)
Comment (rob-bygrave) [reply ebean-api/src/main/java/io/ebean/Database.java:170]: Good thanks.

File: ebean-core/src/main/java/io/ebeaninternal/server/core/DefaultServer.java (hunk 8 above)
Comment (rob-bygrave) [reply ebean-core/src/main/java/io/ebeaninternal/server/core/DefaultServer.java:632]: Behaviour change means it should be in its own PR.  We don't want subtle behaviour changes in a PR that says "Improved Null-checking annotations"  and we do want behaviour changes in their own PR that justifies why the change is good.

**Comment by rPraml** [comment]
Hello @rob-bygrave, @jnehlmeier,

thank you for your constructive feedback.

Rob already merged some changes with #3066

I'll try to summarize, what's left in this PR
- I will revert all code changes here and create extra PRs
- nextId: Yes - it makes no sense here for @Nullable
- currentTransaction: I'm unsure myself here. 
I understand Rob's argument that this could be a negative overall.
However, at least we have pieces of code where we make sure that this is not called with an active transaction.
For example, this is mostly code, that interacts with external APIs, which can't be roll backed.
```java
void deleteUser(String userId) {
   assert DB.currentTransaction() == null : "must not run in transaction";
   if (externalApi.deleteUser(userId)) {
       user = DB.find(User.class, userId);
       user.setDeleted(true);
       DB.save(user);
   }
}
```
It would be bad, if this code runs in a transaction, that is rolled back afterwards.
I can live with a potential IDE warning 'currentTransaction() is never null here' 
- what to do with methods like `boolean delete(Object bean, @Nullable Transaction transaction)`?

I will not continue with the pull request for the time being, but will make it available for discussion first, since it only solves cosmetic problems in the IDE for us.

Roland

**Comment by rbygrave** [comment]
> `assert DB.currentTransaction() == null : "must not run in transaction";`

Ok, I see.  To me I'd say this use case is like `@Transaction(type=NEVER)` but in programatic form.

My gut says that the alternative is for `currentTransaction()` to always return a non nullable transaction and that a new  `boolean hasCurrentTransaction()` be added.  This is probably based on the thought that the most common use of `currentTransaction()` [that I was expecting] is it being used inside methods annotated with `@Transactional` and for these cases the code is always expects non-nullable transaction. 

Hmm, pondering.


> boolean delete(Object bean, @Nullable Transaction transaction)

For these cases I don't see any downside apart from verbosity. I don't think it negatively impacts existing code that is using those methods.

**Comment by rbygrave** [comment]
I think we close this one and have a new fresh PR if desired.

**Pull request closed without merging**